File: backend/app/core/sentry_sanitize.py
"""Sentry before_send sanitization — never ship secrets / PII payloads.

Strips authorization headers, tokens, passwords, payment requisites,
document contents, personal messages, and file bodies from events.
"""
from __future__ import annotations

import re
from typing import Any

SENSITIVE_KEY_RE = re.compile(
    r"(authorization|access[_-]?token|refresh[_-]?token|password|secret|"
    r"api[_-]?key|private[_-]?key|cookie|set-cookie|yookassa|payment[_-]?requisite|"
    r"card[_-]?number|cvv|document[_-]?content|file[_-]?body|message[_-]?body|"
    r"personal[_-]?message|bearer)",
    re.I,
)

VALUE_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"(?i)(bearer\s+)[A-Za-z0-9._\-+=/]+"), r"\1[REDACTED]"),
    (re.compile(r"(?i)(password\s*[=:]\s*)\S+"), r"\1[REDACTED]"),
    (re.compile(r"(?i)(api[_-]?key\s*[=:]\s*)\S+"), r"\1[REDACTED]"),
]


def _redact_string(value: str) -> str:
    out = value
    for pat, repl in VALUE_PATTERNS:
        out = pat.sub(repl, out)
    return out


def redact_mapping(data: Any, *, depth: int = 0) -> Any:
    """Recursively redact sensitive keys/values. Caps depth to avoid cycles."""
    if depth > 8:
        return "[REDACTED_DEPTH]"
    if isinstance(data, dict):
        clean: dict[str, Any] = {}
        for key, val in data.items():
            ks = str(key)
            if SENSITIVE_KEY_RE.search(ks):
                clean[ks] = "[REDACTED]"
            else:
                clean[ks] = redact_mapping(val, depth=depth + 1)
        return clean
    if isinstance(data, list):
        return [redact_mapping(v, depth=depth + 1) for v in data[:50]]
    if isinstance(data, str):
        if len(data) > 2048:
            return "[REDACTED_LARGE_BODY]"
        return _redact_string(data)
    return data


def sentry_before_send(
    event: dict[str, Any], _hint: dict[str, Any] | None = None
) -> dict[str, Any] | None:
    """Sentry SDK before_send hook."""
    if not event:
        return event
    for section in ("request", "extra", "contexts", "user", "tags"):
        if section in event and event[section] is not None:
            event[section] = redact_mapping(event[section])
    req = event.get("request")
    if isinstance(req, dict):
        headers = req.get("headers")
        if isinstance(headers, dict):
            req["headers"] = {
                k: ("[REDACTED]" if SENSITIVE_KEY_RE.search(str(k)) else v)
                for k, v in headers.items()
            }
        if "data" in req:
            req["data"] = redact_mapping(req.get("data"))
        if "cookies" in req:
            req["cookies"] = "[REDACTED]"
        event["request"] = req
    crumbs = event.get("breadcrumbs")
    if isinstance(crumbs, dict) and isinstance(crumbs.get("values"), list):
        for crumb in crumbs["values"]:
            if isinstance(crumb, dict):
                if "data" in crumb:
                    crumb["data"] = redact_mapping(crumb["data"])
                msg = crumb.get("message")
                if isinstance(msg, str) and len(msg) > 512:
                    crumb["message"] = "[REDACTED_LARGE_BODY]"
    return event

File: apps/mobile/lib/sentrySanitize.ts
/**
 * Sentry beforeSend sanitization — strip tokens, auth, passwords,
 * payment requisites, document/message/file bodies. Never log secret values.
 */

const SENSITIVE_KEY =
  /authorization|access[_-]?token|refresh[_-]?token|password|secret|api[_-]?key|private[_-]?key|cookie|yookassa|payment[_-]?requisite|card[_-]?number|cvv|document[_-]?content|file[_-]?body|message[_-]?body|personal[_-]?message|bearer/i;

function redactString(value: string): string {
  let out = value;
  out = out.replace(/(bearer\s+)[A-Za-z0-9._\-+=/]+/gi, '$1[REDACTED]');
  out = out.replace(/(password\s*[=:]\s*)\S+/gi, '$1[REDACTED]');
  if (out.length > 2048) return '[REDACTED_LARGE_BODY]';
  return out;
}

export function redactMapping(data: unknown, depth = 0): unknown {
  if (depth > 8) return '[REDACTED_DEPTH]';
  if (data == null) return data;
  if (Array.isArray(data)) {
    return data.slice(0, 50).map((v) => redactMapping(v, depth + 1));
  }
  if (typeof data === 'object') {
    const src = data as Record<string, unknown>;
    const clean: Record<string, unknown> = {};
    for (const [key, val] of Object.entries(src)) {
      if (SENSITIVE_KEY.test(key)) clean[key] = '[REDACTED]';
      else clean[key] = redactMapping(val, depth + 1);
    }
    return clean;
  }
  if (typeof data === 'string') return redactString(data);
  return data;
}

/** Compatible with @sentry/react-native Event shape (partial). */
export function sanitizeSentryEvent<T extends Record<string, unknown>>(event: T): T {
  const next = { ...event } as Record<string, unknown>;
  for (const section of ['request', 'extra', 'contexts', 'user', 'tags'] as const) {
    if (next[section] != null) {
      next[section] = redactMapping(next[section]);
    }
  }
  const req = next.request as Record<string, unknown> | undefined;
  if (req && typeof req === 'object') {
    if (req.headers && typeof req.headers === 'object') {
      const headers = req.headers as Record<string, unknown>;
      req.headers = Object.fromEntries(
        Object.entries(headers).map(([k, v]) => [k, SENSITIVE_KEY.test(k) ? '[REDACTED]' : v]),
      );
    }
    if ('data' in req) req.data = redactMapping(req.data);
    if ('cookies' in req) req.cookies = '[REDACTED]';
    next.request = req;
  }
  return next as T;
}
